packages/main/src/files/file-manager.ts

```diff
@@ -0,0 +1,164 @@
+import { FilesRequestUrl, getHeaders, makeFilesRequest } from "./request";
+import {
+  FileMetadata,
+  FileMetadataResponse,
+  ListFilesResponse,
+  ListParams,
+  UploadFileResponse,
+} from "./types";
+import { FilesTask } from "./constants";
+
+// Internal type, metadata sent in the upload
+export interface UploadMetadata {
+  name?: string;
+  ["display_name"]?: string;
+}
+
+/**
+ * Class for managing GoogleAI file uploads.
+ * @public
+ */
+export class GoogleAIFileManager {
+  constructor(
+    public apiKey: string,
+    private _requestOptions?: RequestOptions,
+  ) {}
+
+  /**
+   * Upload a file
+   */
+  async uploadFile(
+    filePath: string,
+    fileMetadata: FileMetadata,
+  ): Promise<UploadFileResponse> {
+    const file = readFileSync(filePath);
+    const url = new FilesRequestUrl(
+      FilesTask.UPLOAD,
+      this.apiKey,
+      this._requestOptions,
+    );
+
+    const uploadHeaders = getHeaders(url);
+    const boundary = genBoundary();
+    uploadHeaders.append("X-Goog-Upload-Protocol", "multipart");
+    uploadHeaders.append(
+      "Content-Type",
+      `multipart/related; boundary=${boundary}`,
+    );
+
+    const uploadMetadata: FileMetadata = {
+      mimeType: fileMetadata.mimeType,
+    };
+    uploadMetadata.displayName = fileMetadata.displayName;
+    uploadMetadata.name = fileMetadata.name;
+    if (!uploadMetadata.name?.includes("/")) {
+      uploadMetadata.name = `files/${uploadMetadata.name}`;
+    }
+
+    // Multipart formatting code taken from @firebase/storage
+    const metadataString = JSON.stringify({ file: uploadMetadata });
+    const preBlobPart =
+      "--" +
+      boundary +
+      "\r\n" +
+      "Content-Type: application/json; charset=utf-8\r\n\r\n" +
+      metadataString +
+      "\r\n--" +
+      boundary +
+      "\r\n" +
+      "Content-Type: " +
+      fileMetadata.mimeType +
+      "\r\n\r\n";
+    const postBlobPart = "\r\n--" + boundary + "--";
+    const blob = new Blob([preBlobPart, file, postBlobPart]);
+
+    const response = await makeFilesRequest(url, uploadHeaders, blob);
+    return response.json();
+  }
+
+  /**
+   * List all uploaded files
+   */
+  async listFiles(listParams?: ListParams): Promise<ListFilesResponse> {
+    const url = new FilesRequestUrl(
+      FilesTask.LIST,
+      this.apiKey,
+      this._requestOptions,
+    );
+    if (listParams?.pageSize) {
+      url.appendParam("pageSize", listParams.pageSize.toString());
+    }
+    if (listParams?.pageToken) {
+      url.appendParam("pageToken", listParams.pageToken);
+    }
+    const uploadHeaders = getHeaders(url);
+    const response = await makeFilesRequest(url, uploadHeaders);
+    return response.json();
+  }
+
+  /**
+   * Get metadata for file with given ID
+   */
+  async getFile(fileId: string): Promise<FileMetadataResponse> {
+    const url = new FilesRequestUrl(
+      FilesTask.GET,
+      this.apiKey,
+      this._requestOptions,
+    );
+    url.appendPath(parseFileId(fileId));
```

Comment:
If the `fileId` is just `files/`, `parseFileId` returns an empty string, right? Should we validate the input?

Reply:
Good idea, added some tests as well.